### PR TITLE
chore: use upstream FlightSQL metadata implementation

### PR DESCRIPTION
**Pull request by alamb**
# Rationale

We have upstreamed a bunch of the FlightSQL metadata building code (thanks @roeap!)

https://docs.rs/arrow-flight/41.0.0/arrow_flight/sql/metadata/index.html

(and I am still trying to catch up to @carols10cents 's deletion record)

# Changes

Change IOx to use the upstream implementation

### CODE REVIEW

File: flightsql/src/planner.rs [hunk 1]
```diff
@@ -61,34 +59,28 @@ impl FlightSQLPlanner {
                 get_schema_for_query(handle.query(), ctx).await
             }
             FlightSQLCommand::CommandGetSqlInfo(CommandGetSqlInfo { .. }) => {
-                encode_schema(iox_sql_info_list().schema())
```
Comment (alamb) [inline flightsql/src/planner.rs:64]: encoding the schema can now be easily done as part of creating `FlightInfo`

File: flightsql/src/planner.rs [hunk 2]
```diff
@@ -344,8 +315,17 @@ async fn plan_get_sql_info(ctx: &IOxSessionContext, info: Vec<u32>) -> Result<Lo
...
 
-async fn plan_get_catalogs(ctx: &IOxSessionContext) -> Result<LogicalPlan> {
-    Ok(ctx.batch_to_logical_plan(get_catalogs(ctx.inner())?)?)
+/// Add tables from the DataFusion catalog to the builder
+async fn plan_get_catalogs(
+    ctx: &IOxSessionContext,
+    cmd: CommandGetCatalogs,
+) -> Result<LogicalPlan> {
+    let mut builder = cmd.into_builder();
+    for catalog_name in ctx.inner().catalog_names() {
```
Comment (alamb) [inline flightsql/src/planner.rs:324]: The new upstream builders handle filtering and sorting so the code in IOx no longer needs to do so

File: flightsql/src/planner.rs [hunk 3]
```diff
@@ -361,12 +341,28 @@ async fn plan_get_cross_reference(
...
-    db_schema_filter_pattern: Option<String>,
+    cmd: CommandGetDbSchemas,
 ) -> Result<LogicalPlan> {
-    let batch = get_db_schemas(ctx.inner(), catalog, db_schema_filter_pattern)?;
+    let mut builder = cmd.into_builder();
+    let catalog_list = ctx.inner().state().catalog_list();
+
+    for catalog_name in catalog_list.catalog_names() {
+        // we just got the catalog name from the catalog_list, so it
+        // should always be Some, but avoid unwrap to be safe
```
Comment (NGA-TRAN) [inline flightsql/src/planner.rs:354]: 👍

**Review by NGA-TRAN — APPROVED** [review]
Nice

File: flightsql/src/planner.rs [hunk 4]
```diff
@@ -280,6 +255,7 @@ impl FlightSQLPlanner {
                 let parameter_schema = vec![];
 
                 let dataset_schema = get_schema_for_query(&query, ctx).await?;
+                let dataset_schema = encode_schema(dataset_schema.as_ref())?;
```
Comment (appletreeisyellow) [inline flightsql/src/planner.rs:258]: 👍

**Review by mhilton — APPROVED** [review]

File: flightsql/src/planner.rs [hunk 5]
```diff
@@ -400,27 +396,73 @@ async fn plan_get_primary_keys(
...
+                continue
+            };
+
+            for table_name in schema.table_names() {
+                let Some(table) = schema.table(&table_name).await else {
+                    continue
+                };
+
+                // If the table type doesn't match the types specified in the query, skip
+                let table_type = table_type_name(table.table_type());
```
Comment (appletreeisyellow) [inline flightsql/src/planner.rs:440]: I know `TableType` is an enum that has three values: `Base`, `View`, and `Temporary`. Do you mean the type specified in the query might be something else other than the enum values?

**Review by appletreeisyellow — APPROVED** [review]
Thanks for bringing in the upstream implementation! It looks so neat now! I have one question below just for my understanding

File: flightsql/src/planner.rs (hunk 5 above)
Comment (alamb) [reply flightsql/src/planner.rs:440]: 🤔  I think this comment is left over from a previous implementation and is no longer relevant. I will remove it.

Nice catch @appletreeisyellow

**Pull request merged**